README.rst
Python Client for Cloud Recommender API API
===========================================

|stable| |pypi| |versions|

`Cloud Recommender API API`_: 

- `Client Library Documentation`_
- `Product Documentation`_

.. |stable| image:: https://img.shields.io/badge/support-stable-gold.svg
   :target: https://github.com/googleapis/google-cloud-python/blob/main/README.rst#stability-levels
.. |pypi| image:: https://img.shields.io/pypi/v/google-cloud-recommender.svg
   :target: https://pypi.org/project/google-cloud-recommender/
.. |versions| image:: https://img.shields.io/pypi/pyversions/google-cloud-recommender.svg
   :target: https://pypi.org/project/google-cloud-recommender/
.. _Cloud Recommender API API: https://cloud.google.com/recommender
.. _Client Library Documentation: https://cloud.google.com/python/docs/reference/recommender/latest
.. _Product Documentation:  https://cloud.google.com/recommender

Quick Start
-----------

In order to use this library, you first need to go through the following steps:

1. `Select or create a Cloud Platform project.`_
2. `Enable billing for your project.`_
3. `Enable the Cloud Recommender API API.`_
4. `Setup Authentication.`_

.. _Select or create a Cloud Platform project.: https://console.cloud.google.com/project
.. _Enable billing for your project.: https://cloud.google.com/billing/docs/how-to/modify-project#enable_billing_for_a_project
.. _Enable the Cloud Recommender API API.:  https://cloud.google.com/recommender
.. _Setup Authentication.: https://googleapis.dev/python/google-api-core/latest/auth.html

Installation
~~~~~~~~~~~~

Install this library in a `virtualenv`_ using pip. `virtualenv`_ is a tool to
create isolated Python environments. The basic problem it addresses is one of
dependencies and versions, and indirectly permissions.

With `virtualenv`_, it's possible to install this library without needing system
install permissions, and without clashing with the installed system
dependencies.

.. _`virtualenv`: https://virtualenv.pypa.io/en/latest/


Code samples and snippets
~~~~~~~~~~~~~~~~~~~~~~~~~

Code samples and snippets live in the `samples/` folder.


Supported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^
Our client libraries are compatible with all current [active](https://devguide.python.org/devcycle/#in-development-main-branch) and [maintenance](https://devguide.python.org/devcycle/#maintenance-branches) versions of
Python.

Python >= 3.7

Unsupported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^^^
Python <= 3.6

If you are using an [end-of-life](https://devguide.python.org/devcycle/#end-of-life-branches)
version of Python, we recommend that you update as soon as possible to an actively supported version.


Mac/Linux
^^^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    source <your-env>/bin/activate
    <your-env>/bin/pip install google-cloud-recommender


Windows
^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    <your-env>\Scripts\activate
    <your-env>\Scripts\pip.exe install google-cloud-recommender

Next Steps
~~~~~~~~~~

-  Read the `Client Library Documentation`_ for Cloud Recommender API API
   to see other available methods on the client.
-  Read the `Cloud Recommender API API Product documentation`_ to learn
   more about the product and see How-to Guides.
-  View this `README`_ to see the full list of Cloud
   APIs that we cover.

.. _Cloud Recommender API API Product documentation:  https://cloud.google.com/recommender
.. _README: https://github.com/googleapis/google-cloud-python/blob/main/README.rst


CHANGELOG.md
# Changelog

[PyPI History][1]

[1]: https://pypi.org/project/google-cloud-recommender/#history

## [1.1.3](https://github.com/googleapis/python-recommender/compare/v1.1.2...v1.1.3) (2022-06-08)


### Bug Fixes

* **deps:** require protobuf<4.0.0 on v1 branch ([#212](https://github.com/googleapis/python-recommender/issues/212)) ([f2c31d6](https://github.com/googleapis/python-recommender/commit/f2c31d6202fb86bb25212f3e3d4e7287eec766da))

### [1.1.2](https://github.com/googleapis/python-recommender/compare/v1.1.1...v1.1.2) (2022-03-25)


### Bug Fixes

* **deps:** require google-api-core >= 1.31.5, >= 2.3.2 on v1 release ([#183](https://github.com/googleapis/python-recommender/issues/183)) ([bb07103](https://github.com/googleapis/python-recommender/commit/bb071039cbf27455dd5f4333b853942c3f4ebe34))

### [1.1.1](https://www.github.com/googleapis/python-recommender/compare/v1.1.0...v1.1.1) (2020-10-29)


### Bug Fixes

* tweak retry params for 'ListInsights'/'GetInsight'/'MarkInsightAccepted' API calls (via synth) ([#49](https://www.github.com/googleapis/python-recommender/issues/49)) ([0d2baaf](https://www.github.com/googleapis/python-recommender/commit/0d2baaf9e0d05897b4ea380510d3d899638cb45d))

## [1.1.0](https://www.github.com/googleapis/python-recommender/compare/v1.0.0...v1.1.0) (2020-07-13)


### Features

* add methods for interacting with insights ([#35](https://www.github.com/googleapis/python-recommender/issues/35)) ([940a3fb](https://www.github.com/googleapis/python-recommender/commit/940a3fb01013865c836bfb55397c25284004a7ad))


### Bug Fixes

* update retry config ([#31](https://www.github.com/googleapis/python-recommender/issues/31)) ([5c497e2](https://www.github.com/googleapis/python-recommender/commit/5c497e29d65d288a4b8b24a7b5aa487a5804e880))

## [1.0.0](https://www.github.com/googleapis/python-recommender/compare/v0.3.0...v1.0.0) (2020-05-21)


### Features

* release as production/stable ([#17](https://www.github.com/googleapis/python-recommender/issues/17)) ([b6f0a19](https://www.github.com/googleapis/python-recommender/commit/b6f0a1972df2d0eb49580e152f47b2d13ea1c53c))

## [0.3.0](https://www.github.com/googleapis/python-recommender/compare/v0.2.0...v0.3.0) (2020-03-14)


### Features

* add insight support; undeprecate resource name helper methods (via synth) ([#7](https://www.github.com/googleapis/python-recommender/issues/7)) ([876c383](https://www.github.com/googleapis/python-recommender/commit/876c383afed9a2384d9ef361b4054c381ab9a23b))

## 0.2.0

01-24-2020 14:03 PST

### Implementation Changes
- Deprecate resource name helper methods (via synth).  ([#9863](https://github.com/googleapis/google-cloud-python/pull/9863))

### New Features
- Add v1, set release level to beta. ([#10170](https://github.com/googleapis/google-cloud-python/pull/10170))

### Documentation
- Add Python 2 sunset banner to documentation. ([#9036](https://github.com/googleapis/google-cloud-python/pull/9036))
- Change requests intersphinx url (via synth). ([#9408](https://github.com/googleapis/google-cloud-python/pull/9408))
- Fix library reference doc link. ([#9338](https://github.com/googleapis/google-cloud-python/pull/9338))

### Internal / Testing Changes
- Correct config path in synth file for recommender. ([#10076](https://github.com/googleapis/google-cloud-python/pull/10076))

## 0.1.0

09-27-2019 12:20 PDT

### New Features
- initial release of v1beta1 ([#9257](https://github.com/googleapis/google-cloud-python/pull/9257))


owl-bot-staging/v1beta1/noxfile.py
# -*- coding: utf-8 -*-
# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import os
import pathlib
import shutil
import subprocess
import sys


import nox  # type: ignore

ALL_PYTHON = [
    "3.6",
    "3.7",
    "3.8",
    "3.9",
    "3.10",
]

CURRENT_DIRECTORY = pathlib.Path(__file__).parent.absolute()

LOWER_BOUND_CONSTRAINTS_FILE = CURRENT_DIRECTORY / "constraints.txt"
PACKAGE_NAME = subprocess.check_output([sys.executable, "setup.py", "--name"], encoding="utf-8")

BLACK_VERSION = "black==19.10b0"
BLACK_PATHS = ["docs", "google", "tests", "samples", "noxfile.py", "setup.py"]
DEFAULT_PYTHON_VERSION = "3.9"

nox.sessions = [
    "unit",
    "cover",
    "mypy",
    "check_lower_bounds"
    # exclude update_lower_bounds from default
    "docs",
    "blacken",
    "lint",
    "lint_setup_py",
]

@nox.session(python=ALL_PYTHON)
def unit(session):
    """Run the unit test suite."""

    session.install('coverage', 'pytest', 'pytest-cov', 'pytest-asyncio', 'asyncmock; python_version < "3.8"')
    session.install('-e', '.')

    session.run(
        'py.test',
        '--quiet',
        '--cov=google/cloud/recommender_v1beta1/',
        '--cov=tests/',
        '--cov-config=.coveragerc',
        '--cov-report=term',
        '--cov-report=html',
        os.path.join('tests', 'unit', ''.join(session.posargs))
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def cover(session):
    """Run the final coverage report.
    This outputs the coverage report aggregating coverage from the unit
    test runs (not system test runs), and then erases coverage data.
    """
    session.install("coverage", "pytest-cov")
    session.run("coverage", "report", "--show-missing", "--fail-under=100")

    session.run("coverage", "erase")


@nox.session(python=ALL_PYTHON)
def mypy(session):
    """Run the type checker."""
    session.install('mypy', 'types-pkg_resources')
    session.install('.')
    session.run(
        'mypy',
        '--explicit-package-bases',
        'google',
    )


@nox.session
def update_lower_bounds(session):
    """Update lower bounds in constraints.txt to match setup.py"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'update',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )


@nox.session
def check_lower_bounds(session):
    """Check lower bounds in setup.py are reflected in constraints file"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'check',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )

@nox.session(python=DEFAULT_PYTHON_VERSION)
def docs(session):
    """Build the docs for this library."""

    session.install("-e", ".")
    session.install("sphinx==4.0.1", "alabaster", "recommonmark")

    shutil.rmtree(os.path.join("docs", "_build"), ignore_errors=True)
    session.run(
        "sphinx-build",
        "-W",  # warnings as errors
        "-T",  # show full traceback on exception
        "-N",  # no colors
        "-b",
        "html",
        "-d",
        os.path.join("docs", "_build", "doctrees", ""),
        os.path.join("docs", ""),
        os.path.join("docs", "_build", "html", ""),
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint(session):
    """Run linters.

    Returns a failure if the linters find linting errors or sufficiently
    serious code quality issues.
    """
    session.install("flake8", BLACK_VERSION)
    session.run(
        "black",
        "--check",
        *BLACK_PATHS,
    )
    session.run("flake8", "google", "tests", "samples")


@nox.session(python=DEFAULT_PYTHON_VERSION)
def blacken(session):
    """Run black. Format code to uniform standard."""
    session.install(BLACK_VERSION)
    session.run(
        "black",
        *BLACK_PATHS,
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint_setup_py(session):
    """Verify that setup.py is valid (including RST check)."""
    session.install("docutils", "pygments")
    session.run("python", "setup.py", "check", "--restructuredtext", "--strict")
